Preferred write partitioning threshold

skrzypo987 opened the pull request.

skrzypo987 commented on core/trino-main/src/main/java/io/trino/sql/planner/PlanOptimizers.java:
```diff
@@ -430,6 +431,7 @@ public PlanOptimizers(
                                 .addAll(columnPruningRules)
                                 .addAll(projectionPushdownRules)
                                 .addAll(ImmutableSet.of(
+                                        new PreferWritePartitioning(),
```
To be honest I have no idea where to put it here, so it landed at the beginning.

findepi replied:
@martint suggestions?

sopel39 replied:
Move it before or after reorder joins is applied. The assumption is that for reorder joins we should have the most accurate stats:
* `PredicatePushDown` has been run with `useTableProperties`
* later in the planning exchange nodes are added which might confuse stats.

Please add a comment why there.

skrzypo987 commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/PreferWritePartitioning.java:
```diff
@@ -0,0 +1,59 @@
+            return Result.empty();
+        }
+
+        double numberOfPartitions = node.getPreferredPartitioningScheme().get().getPartitioning().getColumns().stream()
```
Calculating number of partition is non-trivial for multiple partitions.
It is somewhere between max cardinality of one of partitions and product of cardinality of all of them.

skrzypo987 commented:
Docs still missing

tooptoop4 commented:
@skrzypo987 does this fix https://github.com/trinodb/trino/issues/6570 ?

skrzypo987 commented:
> @skrzypo987 does this fix #6570 ?

I am afraid if enabling `use_preferred_write_partitioning` does not work then this will not work either

sopel39 commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/PreferWritePartitioning.java:
```diff
@@ -0,0 +1,50 @@
+import static io.trino.SystemSessionProperties.isUsePreferredWritePartitioning;
+import static io.trino.sql.planner.plan.Patterns.tableWriterNode;
+
+public class PreferWritePartitioning
```
Please add a basic rule test, see `io.trino.sql.planner.iterative.rule.TestApplyTableScanRedirection` how we mock table handles in rule tests.

sopel39 commented on core/trino-main/src/main/java/io/trino/sql/planner/plan/TableWriterNode.java:
```diff
@@ -84,6 +86,7 @@ public TableWriterNode(
         this.columnNames = ImmutableList.copyOf(columnNames);
         this.notNullColumnSymbols = ImmutableSet.copyOf(requireNonNull(notNullColumnSymbols, "notNullColumns is null"));
         this.partitioningScheme = requireNonNull(partitioningScheme, "partitioningScheme is null");
+        this.preferredPartitioningScheme = requireNonNull(preferredPartitioningScheme, "preferredPartitioningScheme is null");
```
add a validation that just one of `partitioningScheme` and `preferredPartitioningScheme` is set at same time

sopel39 commented on core/trino-main/src/test/java/io/trino/sql/planner/TestInsert.java:
```diff
@@ -180,24 +180,22 @@ public void testCreateTableAsSelectUnpartitioned()
                 .build();
     }
 
-    private Session withoutPreferredPartitioning()
```
Let's keep explicit session as it was. It's easier for reader to understand what's going on.

sopel39 commented on core/trino-main/src/test/java/io/trino/sql/planner/TestInsert.java:
```diff
@@ -180,24 +180,22 @@ public void testCreateTableAsSelectUnpartitioned()
+    private Session defaultSession()
     {
         return Session.builder(getQueryRunner().getDefaultSession())
                 .setSystemProperty(USE_PREFERRED_WRITE_PARTITIONING, "false")
```
if it's default, why we change properties?

sopel39 commented on core/trino-main/src/test/java/io/trino/sql/planner/TestInsert.java:
```diff
@@ -99,7 +99,7 @@ public void testInsertWithPreferredPartitioning()
         assertDistributedPlan(
                 "INSERT into test_table_preferred_partitioning VALUES (1, 2)",
-                withPreferredPartitioning(),
+                withPreferredPartitioning(defaultSession()),
```
This adds unnecessary complexity to building of session. IMO previous code was cleaner

sopel39 commented on core/trino-main/src/test/java/io/trino/sql/planner/TestInsert.java:
```diff
@@ -99,7 +99,7 @@ public void testInsertWithPreferredPartitioning()
     {
```
I don't think this commit brings much benefit. I would skip it.

sopel39 commented on core/trino-main/src/main/java/io/trino/SystemSessionProperties.java:
```diff
@@ -209,6 +210,11 @@ public SystemSessionProperties(
                         false),
+                integerProperty(
+                        PREFERRED_WRITE_PARTITIONING_THRESHOLD,
+                        "PreferredWritePartitioningThreshold",
```
fix description

sopel39 commented on core/trino-main/src/main/java/io/trino/sql/analyzer/FeaturesConfig.java:
```diff
@@ -373,6 +374,19 @@ public FeaturesConfig setUsePreferredWritePartitioning(boolean usePreferredWrite
+        return preferredWritePartitioningThreshold;
+    }
+
+    @Config("preferred-write-partitioning-threshold")
```
add description

sopel39 commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/PreferWritePartitioning.java:
```diff
@@ -33,6 +47,14 @@ public Result apply(TableWriterNode node, Captures captures, Context context)
 
+        double numberOfPartitions = node.getPreferredPartitioningScheme().get().getPartitioning().getColumns().stream()
+                .mapToDouble(column -> context.getStatsProvider().getStats(node.getSource()).getSymbolStatistics().get(column).getDistinctValuesCount())
+                .max()
```
IMO that should be multiplication. You could reuse: `io.trino.cost.AggregationStatsRule#groupBy`

sopel39 commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/PreferWritePartitioning.java:
```diff
@@ -33,6 +47,14 @@ public Result apply(TableWriterNode node, Captures captures, Context context)
+                .mapToDouble(column -> context.getStatsProvider().getStats(node.getSource()).getSymbolStatistics().get(column).getDistinctValuesCount())
+                .max()
+                .orElse(1);
+        if (numberOfPartitions < getPreferredWritePartitioningThreshold(context.getSession())) {
```
This should be other way around. If there is lots of partitions, preferred write partitioning should be used.

sopel39 commented on core/trino-main/src/test/java/io/trino/sql/planner/TestInsert.java:
```diff
@@ -188,13 +189,42 @@ public void testCreateTableAsSelectUnpartitioned()
     }
 
+    @Test
+    public void testCreateTableAsSelectWithPreferredPartitioningThreshold()
```
this also deserves rule testing with stats, see `io.trino.sql.planner.iterative.rule.TestReorderJoins` for example

skrzypo987 commented:
I noticed that `use-preferred-write-partitioning` is not mentioned in the docs, just in release notes. Should this follow the same pattern?

sopel39 commented on core/trino-main/src/test/java/io/trino/sql/planner/iterative/rule/TestPreferWritePartitioning.java:
```diff
@@ -0,0 +1,95 @@
+            .setSystemProperty(USE_PREFERRED_WRITE_PARTITIONING, "true")
+            .build();
+
+    private static final PlanMatchPattern MATCH_PATTERN = tableWriter(
```
it's used once. Inline it in test

skrzypo987 replied:
It is used twice. Second usage is in the second commit

sopel39 commented on core/trino-main/src/main/java/io/trino/SystemSessionProperties.java:
```diff
@@ -211,6 +212,11 @@ public SystemSessionProperties(
                         false),
+                integerProperty(
+                        PREFERRED_WRITE_PARTITIONING_THRESHOLD,
+                        "Preferred write partitioning threshold",
```
Rephase it to something like:
```
Use preferred insert partitioning when number of written partitions exceeds given threshold`
```

sopel39 commented on core/trino-main/src/main/java/io/trino/SystemSessionProperties.java:
```diff
@@ -69,6 +69,7 @@
     public static final String QUERY_MAX_STAGE_COUNT = "query_max_stage_count";
     public static final String REDISTRIBUTE_WRITES = "redistribute_writes";
     public static final String USE_PREFERRED_WRITE_PARTITIONING = "use_preferred_write_partitioning";
+    public static final String PREFERRED_WRITE_PARTITIONING_THRESHOLD = "preferred_write_partitioning_threshold";
```
maybe `preferred_write_partitioning_min_number_of_partitions`

sopel39 commented on core/trino-main/src/test/java/io/trino/sql/planner/iterative/rule/TestPreferWritePartitioning.java:
```diff
@@ -93,6 +104,52 @@ public void testDoesNotFire()
+                ImmutableList.of(new Symbol("col_one"), new Symbol("col_two")))))
+                .withSession(SESSION_WITH_PREFERRED_PARTITIONING_THRESHOLD)
+                .overrideStats("mock", stats)
+                .matches(MATCH_PATTERN);
```
change name of `MATCH_PATTERN` to be more meaningful

skrzypo987 replied:
Changed to `SUCCESSFUL_MATCH`.

sopel39 commented on core/trino-main/src/test/java/io/trino/sql/planner/iterative/rule/TestPreferWritePartitioning.java:
```diff
@@ -93,6 +104,52 @@ public void testDoesNotFire()
     }
 
+    @Test
+    public void testThreshold()
```
What does it test? Why it did not fire?

skrzypo987 replied:
It fires without a threshold, so when the threshold `2` is added it should not fire since there is only one partition

sopel39 replied:
please add a comment

findepi commented on core/trino-main/src/main/java/io/trino/SystemSessionProperties.java:
```diff
@@ -211,6 +212,11 @@ public SystemSessionProperties(
                         false),
+                integerProperty(
+                        PREFERRED_WRITE_PARTITIONING_MIN_NUMBER_OF_PARTITIONS,
+                        "Use preferred write partitioning when number of written partitions exceeds given threshold`",
```
Would it be possible to make the configuration toggle a little bit automatic?
I.e. how is the user going to determine whether they want to set this? 
If this depends e.g. on number of partitions vs number of nodes, we could encode this reasoning within server's logic

skrzypo987 replied:
The idea was to leave it backward-compatible (`0` as the default value) and let the user decide the threshold based on his/her workload if necessary.
However making this more "intelligent" is not a bad idea. The question is whether we want to do it here or as a follow-up.

sopel39 replied:
`when number of written` or `when the number of written`. I'm not native, but I think the latter is correct

sopel39 replied:
> If this depends e.g. on number of partitions vs number of nodes, we could encode this reasoning within server's logic

This doesn't really depend on number of workers. The hard limit is `hive.max-partitions-per-writers` which is independent of workers.

sopel39 commented on core/trino-main/src/main/java/io/trino/sql/analyzer/FeaturesConfig.java:
```diff
@@ -80,6 +80,7 @@
     private int maxReorderedJoins = 9;
     private boolean redistributeWrites = true;
     private boolean usePreferredWritePartitioning;
+    private int preferredWritePartitioningMinNumberOfPartitions;
```
I suggest to set it to some default value, e.g: `50`. This should:
* make preferred partitioning to be used when there is full table CTAS with more than 50 partitions
* provide good concurrency (>50) for preferred partitioning
* mitigate hive writer partition limit (100)
* provide good concurrency for inserts into single partition

WDYT: @findepi @willmostly ?

sopel39 replied:
Please set it to 50.

sopel39 commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/PreferWritePartitioning.java:
```diff
@@ -46,6 +48,18 @@ public Result apply(TableWriterNode node, Captures captures, Context context)
         }
 
+        double numberOfPartitions = node.getPreferredPartitioningScheme().get().getPartitioning().getColumns().stream()
+                .mapToDouble(column ->
```
Please extract and reuse code from `io.trino.cost.AggregationStatsRule`. When we change computations for aggregation stats, we want to cover this place too.

sopel39 commented:
> I noticed that use-preferred-write-partitioning is not mentioned in the docs, just in release notes. Should this follow the same pattern?

Nope. Both should be documented.

skrzypo987 commented:
Added docs and updated the code.
Things still pending:

- Default threshold value
- Where to put the `PreferWritePartitioning` class in the optimizer (or leave it as it is)

sopel39 commented on docs/src/main/sphinx/admin/properties-writer-scaling.rst:
```diff
@@ -33,3 +33,11 @@ another writer is eligible to be added. Each writer task may have multiple
+* **Type:** ``boolean``
+* **Default value:** ``false``
+
+Enables preferred write partitioning. This option will force every partition
```
this suggest that all partitions will be written by single worker, which is not the case.
I suggest to extract this commit as separate PR and tag @mosabua

tooptoop4 replied:
that is a needed feature when writing small dataset!

sopel39 commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/PreferWritePartitioning.java:
```diff
@@ -46,6 +48,14 @@ public Result apply(TableWriterNode node, Captures captures, Context context)
+                context.getStatsProvider().getStats(node.getSource()),
+                node.getPreferredPartitioningScheme().get().getPartitioning().getColumns());
+
+        if (numberOfPartitions < getPreferredWritePartitioningThreshold(context.getSession())) {
```
Make it <=. This way we could always turn it on (by setting value to `0`).

sopel39 replied:
make this `<=`. This way we can always turn it on (even for single partition) even if stats are off.

sopel39 commented on core/trino-main/src/test/java/io/trino/sql/planner/TestInsert.java:
```diff
@@ -197,6 +220,16 @@ private Session withPreferredPartitioning()
+        return Session.builder(getQueryRunner().getDefaultSession())
+                .setSystemProperty(USE_PREFERRED_WRITE_PARTITIONING, "true")
+                .setSystemProperty(PREFERRED_WRITE_PARTITIONING_MIN_NUMBER_OF_PARTITIONS, "2")
+                .setSystemProperty(REDISTRIBUTE_WRITES, "false")
```
We should use default value for `REDISTRIBUTE_WRITES`, see: https://github.com/trinodb/trino/pull/7034

sopel39 commented on docs/src/main/sphinx/admin/properties-writer-scaling.rst:
```diff
@@ -35,9 +35,17 @@ divided by the number of writers per task. This can be specified on a
+* **Default value:** ``0``
+
+The minimum number of partitions required to trigger the ``preferred write
+partitioning`` mechanism.
```
`mechansm -> optimization`

Elaborate more what `partitions` it refers to.

sopel39 commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/PreferWritePartitioning.java:
```diff
@@ -46,6 +48,14 @@ public Result apply(TableWriterNode node, Captures captures, Context context)
             return Result.empty();
         }
 
+        double numberOfPartitions = getRowsCount(
```
rename to `expectedNumberOfPartitions`

sopel39 commented on core/trino-main/src/test/java/io/trino/sql/planner/iterative/rule/TestPreferWritePartitioning.java:
```diff
@@ -93,6 +104,53 @@ public void testDoesNotFire()
+    }
+
+    @Test
+    public void testThresholdWithMultiplePartitions()
```
Add a test with missing or `NaN` stats. When stats are missing, we should fallback to not use preferred partitioning. However, it should be possible to force using preferred partitioning (e.g threshold set to 0). This should be documented

sopel39 commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/PreferWritePartitioning.java:
```diff
@@ -0,0 +1,73 @@
+    @Override
+    public boolean isEnabled(Session session)
+    {
+        return isUsePreferredWritePartitioning(session);
```
we should also enable `use-preferred-write-partitioning` by default.

skrzypo987 commented:
Another version.
Handles no stats/nan case gracefully. See `TestPreferWritePartitioning` for a list of cases.
Docs updated, however waiting for https://github.com/trinodb/trino/pull/7043 to be finally polished.
There is, however, a problem with naming - `min number of partitions` implies `<` logic, but it actually is `<=`. At this point we can use `<` since `0` is explicitly handled and keep the `min` in the name.
@sopel39

sopel39 commented on docs/src/main/sphinx/admin/properties-writer-scaling.rst:
```diff
@@ -33,3 +33,11 @@ another writer is eligible to be added. Each writer task may have multiple
 divided by the number of writers per task. This can be specified on a
 per-query basis using the ``writer_min_size`` session property.
+
+``use-preferred-write-partitioning``
```
remove this commit? There is a separete PR already

sopel39 commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/PreferWritePartitioning.java:
```diff
@@ -46,6 +49,28 @@ public Result apply(TableWriterNode node, Captures captures, Context context)
         }
 
+        int minimumNumberOfPartitions = getPreferredWritePartitioningMinNumberOfPartitions(context.getSession());
+        if (minimumNumberOfPartitions == 0) {
```
Make it `<= 1`. I would say if somebody sets min to `1`, then preferred partitioning should always be used since one could pretty much assume insert will write to at least one partition

sopel39 commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/PreferWritePartitioning.java:
```diff
@@ -46,6 +49,28 @@ public Result apply(TableWriterNode node, Captures captures, Context context)
+                context.getStatsProvider().getStats(node.getSource()),
+                node.getPreferredPartitioningScheme().get().getPartitioning().getColumns());
+
+        if (isNaN(expectedNumberOfPartitions) || expectedNumberOfPartitions == 0) {
```
`expectedNumberOfPartitions == 0` doesn't seem to be needed.

sopel39 commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/PreferWritePartitioning.java:
```diff
@@ -46,6 +49,28 @@ public Result apply(TableWriterNode node, Captures captures, Context context)
+            return Result.empty();
+        }
+
+        if (expectedNumberOfPartitions <= minimumNumberOfPartitions) {
```
This should be `<`. When expected number of partitions >= min, then we should use preferred partitioning

sopel39 commented on core/trino-main/src/test/java/io/trino/sql/planner/iterative/rule/TestPreferWritePartitioning.java:
```diff
@@ -71,38 +86,93 @@ public void tearDown()
+    @Test(dataProvider = "preferWritePartitioningDataProvider")
+    public void testPreferWritePartitioning(Session session, double distinctValuesStat, boolean match)
+    {
+        PlanNodeStatsEstimate stats = PlanNodeStatsEstimate.builder()
```
generate `PlanNodeStatsEstimate` in `if (distinctValuesStat != NO_STATS) {` branch

sopel39 commented on core/trino-main/src/test/java/io/trino/sql/planner/iterative/rule/TestPreferWritePartitioning.java:
```diff
@@ -71,38 +86,93 @@ public void tearDown()
-                        Partitioning.create(FIXED_HASH_DISTRIBUTION, ImmutableList.of()),
-                        ImmutableList.of())))
-                .withSession(SESSION_WITH_PREFERRED_PARTITIONING)
+        PlanNodeStatsEstimate stats = PlanNodeStatsEstimate.builder()
```
add a comment what does it test exactly

sopel39 commented on docs/src/main/sphinx/admin/properties-writer-scaling.rst:
```diff
@@ -35,9 +35,20 @@ divided by the number of writers per task. This can be specified on a
 Enables preferred write partitioning. This option will force every partition
 to be written by a single worker.
+
+``preferred-write-partitioning-min-number-of-partitions``
```
follow up PR. I would like @mosabua to review it.

skrzypo987 replied:
Maybe a follow-up commit? Features and docs should IMO be added atomically.

sopel39 commented on docs/src/main/sphinx/admin/properties-writer-scaling.rst:
```diff
@@ -35,9 +35,20 @@ divided by the number of writers per task. This can be specified on a
+* **Default value:** ``50``
+
+The minimum number of partitions required to trigger the ``preferred write
+partitioning`` optimisation. If the number of partitions cannot be estimated
```
`if the number of partitions cannot be estimated or preferred-write-partitioning-min-number-of-partitions > 1 then preferred write partitioning won't be used. Setting preferred-write-partitioning-min-number-of-partitions to value less or equal to 1 will force Trino to use preferrered write partitioning`

skrzypo987 replied:
`or preferred-write-partitioning-min-number-of-partitions > 1 then preferred write partitioning won't be used` means that setting this value > 1 will always disable the feature. And this is not the case.
Let's wait for @mosabua to comment on that.

sopel39 commented:
> There is, however, a problem with naming - min number of partitions implies < logic, but it actually is <=. At this point we can use < since 0 is explicitly handled and keep the min in the name.

I understand it that if estimated number of partitions >= min threshold, we use preferred write partitioning.

sopel39 commented on core/trino-main/src/main/java/io/trino/sql/analyzer/FeaturesConfig.java:
```diff
@@ -373,6 +374,20 @@ public FeaturesConfig setUsePreferredWritePartitioning(boolean usePreferredWrite
+    }
+
+    @Config("preferred-write-partitioning-min-number-of-partitions")
+    @ConfigDescription("Minimal number of partitions to enable preferred write partitioning")
```
make session and feature descriptions consistent

sopel39 commented on core/trino-main/src/main/java/io/trino/SystemSessionProperties.java:
```diff
@@ -209,6 +210,11 @@ public SystemSessionProperties(
                         false),
+                integerProperty(
+                        PREFERRED_WRITE_PARTITIONING_MIN_NUMBER_OF_PARTITIONS,
+                        "Minimum number of partitions to enable preferred write partitioning`",
```
Actually, I think the other comment was better:
```
Use preferred write partitioning when the number of written partitions exceeds given threshold
```

mosabua replied:
@sopel39 s version but 

exceeds the configured threshold

sopel39 commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/PreferWritePartitioning.java:
```diff
@@ -46,6 +49,28 @@ public Result apply(TableWriterNode node, Captures captures, Context context)
             return Result.empty();
         }
 
+        int minimumNumberOfPartitions = getPreferredWritePartitioningMinNumberOfPartitions(context.getSession());
```
add a comment that this allows to use preferred write partitioning even if stats are missing or broken

sopel39 commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/PreferWritePartitioning.java:
```diff
@@ -46,6 +49,28 @@ public Result apply(TableWriterNode node, Captures captures, Context context)
+                context.getStatsProvider().getStats(node.getSource()),
+                node.getPreferredPartitioningScheme().get().getPartitioning().getColumns());
+
+        if (isNaN(expectedNumberOfPartitions)) {
```
make a single iff of `isNaN(expectedNumberOfPartitions)` and `expectedNumberOfPartitions < minimumNumberOfPartitions`

sopel39 commented on core/trino-main/src/test/java/io/trino/sql/planner/TestInsert.java:
```diff
@@ -166,7 +167,7 @@ public void testCreateTableAsSelectWithRequiredPartitioning()
         assertDistributedPlan(
                 "CREATE TABLE new_test_table_required_partitioning (column1, column2) AS SELECT * FROM (VALUES (1, 2)) t(column1, column2)",
-                withPreferredPartitioning(),
+                withForcedPreferredPartitioning(),
```
please rebase your code

sopel39 commented on core/trino-main/src/test/java/io/trino/sql/planner/TestInsert.java:
```diff
@@ -187,10 +188,42 @@ public void testCreateTableAsSelectUnpartitioned()
+    {
+        return Session.builder(getQueryRunner().getDefaultSession())
+                .setSystemProperty(USE_PREFERRED_WRITE_PARTITIONING, "true")
+                .setSystemProperty(PREFERRED_WRITE_PARTITIONING_MIN_NUMBER_OF_PARTITIONS, "0")
```
use `1` instead of `0`

sopel39 commented on core/trino-main/src/test/java/io/trino/sql/planner/TestInsert.java:
```diff
@@ -187,10 +188,42 @@ public void testCreateTableAsSelectUnpartitioned()
                                         values("column1", "column2")))));
```
`testCreateTableAsSelectUnpartitioned` should use `withPreferredPartitioning`. Please extract as a separate commit

sopel39 commented on core/trino-main/src/test/java/io/trino/sql/planner/iterative/rule/TestPreferWritePartitioning.java:
```diff
@@ -14,38 +14,53 @@
+            .setCatalog(MOCK_CATALOG)
+            .setSchema(TEST_SCHEMA)
+            .setSystemProperty(USE_PREFERRED_WRITE_PARTITIONING, "true")
+            .setSystemProperty(PREFERRED_WRITE_PARTITIONING_MIN_NUMBER_OF_PARTITIONS, "0")
```
use `1` instead

sopel39 reviewed (commented):
lgtm % comments

sopel39 commented on core/trino-main/src/main/java/io/trino/sql/analyzer/FeaturesConfig.java:
```diff
@@ -80,6 +80,7 @@
     private int maxReorderedJoins = 9;
     private boolean redistributeWrites = true;
     private boolean usePreferredWritePartitioning;
```
add a separate commit that enables `usePreferredWritePartitioning` by default. I think we should atomically change semantics (instead of two PRs) and describe changes in release notes.

sopel39 reviewed (commented):
lgtm % comment about rule placing, let's enable preferred partitioning by default

sopel39 commented on core/trino-main/src/main/java/io/trino/sql/planner/PlanOptimizers.java:
```diff
@@ -695,7 +694,13 @@ public PlanOptimizers(
                         estimatedExchangesCostCalculator,
                         ImmutableSet.of(new RemoveRedundantIdentityProjections())),
-
+                // Prefer write partitioning requires the most accurate stats possible i.e.
```
`Prefer write partitioning` ->
```
Prefer write partitioning rule requires accurate stats.
Therefore PredicatePushDown, columnPruningOptimizer and
RemoveRedundantIdentityProjections need to run beforehand.
```

sopel39 reviewed (approved):
small comment

sopel39 commented on core/trino-main/src/main/java/io/trino/sql/planner/PlanOptimizers.java:
```diff
@@ -693,7 +694,14 @@ public PlanOptimizers(
-
+                // Prefer write partitioning rule requires accurate stats.
+                // Therefore PredicatePushDown, columnPruningOptimizer and
+                // RemoveRedundantIdentityProjections need to run beforehand.
```
nit: we could remove `RemoveRedundantIdentityProjections` from both comments (here and below) as it should not affect stats. This could be separate PR

skrzypo987 replied:
I'll handle it when I get some free cycles
